shared/src/commonMain/kotlin/io/customer/shared/work/JobExecutor.kt

```diff
@@ -0,0 +1,111 @@
+package io.customer.shared.work
+
+import io.customer.shared.util.Dispatcher
+import io.customer.shared.util.Logger
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+
+/**
+ * Helper class to simplify running operations on coroutines and make exception handling easier.
+ * Every class should have its own instance of [JobExecutor] to avoid sharing coroutines with other
+ * classes.
+ */
+internal interface JobExecutor {
```

Comment:
Coroutines error handling is a topic that I am sure has lots of different opinions. I remember when I used RxJava in the past, the same differing opinions occurred in regards to how to handle errors in that library. 

I have been trying to find the article that described my opinion best, but can't seem to find it so I'll try my best to describe it here. 

This PR has a lot of `@Throws` annotations and `kotlin.runCatching{}` code. Making me think that our code has many different places where exceptions can be thrown. Because there are many places that exceptions could be thrown, `JobExecutor` and `JobDispatcher` is introduced into the codebase to make exception catching in coroutines customizable. 

I have built apps in the past that introduced concepts like this where every class or every function could behave differently depending on if there was an exception or not inside of a coroutine. Through this experience, I ended up encountering code that was more complex then it needed to be and when a crash happened in my app, the crashes were harder to debug (since the exceptions propagated up the stack). 

Then, in future apps I built, I changed the design of exception handling to try and encapsulate exception throwing and catching to as little code as possible and instead use the `kotlin.Result<>` data type more. I found this to be a more enjoyable experience. If you are experienced in Rx, an easy way to communicate it is to limit the number of places you have `onError()` callbacks in your code. You can quite easily have `onError()` callbacks exist every time that you launch a coroutine/Rx observable. But, there are drawbacks to that and that's what I am bringing up. 

From reading this PR, it seems that exception throwing is mostly happening in our database layer from the DB library that we use. I bet the networking library we use also can throw exceptions. That limits the number of places that we need to catch and process exceptions 👍. 

Instead of having Coroutine exception catching and handling logic scattered across many class/functions/files in our code, what if we instead encapsulated that exception catching and handling into 1 place and converted that exception thrown into a `kotlin.Result<>` instead? 

So, let's say that that we make 1 class where all DB operations exist:

```kt
class DatabaseQuery {
  fun suspend select(): Result<> {
    // because we are using `suspend`, we can use try/catch. 
    // or, we can use kotlin.runCatching. 
    try {
	// sql delight code exists in here to perform the query
    } catch {
       // we encapsulate exception throwing into this 1 function and convert that exception throw into a Result object that can be parsed later by our SDK code. 
       // This removes the need for coroutines to handle exception handling!
       return Result.failure()
    }
  }
```

This class might be a bad example as it may not be a good idea to introduce a class that performs SELECT queries inside of it via a `select` function. Maybe a better example is with our networking layer of our code. Have all `try/catch` exception catching for exceptions that could be thrown by ktor existing in our HTTP client class. 1 class where all exception catching is done by our networking layer. In the HTTP client, catch the exception and return instead return a `Result.failure` if an exception is thrown. 

By doing this, our coroutines in our code base can be simplified and make the code easier to read. 

- `suspend` functions that call the database or networking code doesn't need exception handling (try/catch, kotlin.runblocking), but instead have code that can parse `Result` objects. 
- Coroutines no longer need onError callbacks as exceptions are no longer being thrown. 

Therefore, I don't currently see the need for `JobExecutor` or `JobDispatcher`. Or at least, the exception handling parts of these classes (maybe there are other features they introduce). 

---

In conclusion, this code review suggestion is not trying to focus on 1 particular piece of code but instead feedback on the entire PR and the design of exception handling. I believe that introducing coroutine exception handling as a feature will make our code harder to read, write, maintain and to debug if there is an error. 

The articles I have read in the past that describe this better then I do, but they try to suggest that if a coroutine catches an exception, it's suggesting you have a bug in your code because your code  should have all exceptions caught directly using try/catch blocks so a couroutine never catches an exception itself. 

I believe that Coroutines are really convenient for moving between threads and write async code. But coroutines are really bad at exception handling. Our code's functions can be really good at exception handling and take exception handling off of coroutine's job.

Reply:
Thanks for sharing your thoughts on this Levi. I do want to discuss this more because I know it is not yet in very refined state, but I think it can add some value. I agree to most of the points you mentioned, but would still like to share my thinking behind it.

I wanted to make it easier to switch to `background` or `main` thread without worrying about creation of coroutines. This is because we put the task from BQ in background, the link breaks. Queue resumes on background and the timer on main. And `JobDispatcher` makes it very clean by providing `runOnBackground` and `runOnMain`.

I couldn't find good article or official docs that could answer my question that should we create new coroutine every time we need, or should we reuse. Because every time our timer runs out, we need a coroutine. This means a new coroutine every 30 seconds (max). So I thought it might be a good idea to reuse same coroutine for one class and `JobExecutor` makes this easier.

But reusing same coroutine can create different challenges. e.g. if a `Job` fails in the coroutine, the coroutine will stop and all jobs in it will be skipped. So I added some `CoroutineExceptionHandler` to it that creates new scope so that new jobs can continue running.

`onError` and `onFailure` methods are optional that were planned to help in cases where we need some useful logs for failure or release locks. But, if we are able to remove locks (which is already in [our discussion](https://github.com/customerio/mobile-shared/pull/5/files#r1054402542)) and assure all callers take responsibility of running safe code in the block and/or not expect the executor to send error callbacks, we can eliminate both of these methods. Which, I think is doable as this is to be used by our SDK only.

I agree we should not expect our code to crash and make proper fixes as needed. But practically, human errors can sometimes lead to crashes. And since we are developing an SDK, I just wanted to add protection layer that could fail silently and not crash the apps. I know this leads to another important discussion that failing silently is not always good as it may result in ignoring important bugs sometimes. But this exactly is what I added `fatal` logs for.

`fatal` can be considered exceptions/errors that can crash app and/or break our SDK. We just catch them so it doesn't break customer's app. To get reports on it, we can probably have an api to report on dashboard, so we can keep an eye on it. I haven't decided anything on this yet as this needs more input from squad. But this was the main idea behind `fatal` logs. And any `fatal` log reported should be treated with high priority as we would want these logs to be empty always and not expect anything there.

For using the `Result` object, I would also prefer in most cases. In fact, I have tried to use it as much as possible. But I'll take a look at code again and see how can we use it more. Meanwhile, if you find any area which can be replaced with `Result`, please do tag me so I can update it accordingly.

I'm sorry if I missed any of your point. I really appreciate this and think this is a very helpful discussion and will definitely lead us to better solutions.

Reply:
Quick update on this, I have removed `onFailure` to make it slightly simpler. We don't expect our code to crash anywhere, but this will still act like 2 step exception handler so if some unfortunate change results in stopping the coroutine, `JobExecutor` will attempt to fix it by restarting the scope.

I would suggest to keep it as it is for now, we'll see how it works and may come back to it later after more people from squad have played around with KMM.

Reply:
I have thought about this conversation quite a bit and I still strongly recommend the [suggestions in my original comment](https://github.com/customerio/mobile-shared/pull/5/files#r1053627678) of making coroutine exception handling simple and not using a coroutine error handler *unless there is an edge case that makes sense* to use it for. 

I appreciate removing something and making it slightly simpler. I am trying to find in the code what was changed but can't easily figure out how to view those changes. Can you help point me to some commits maybe that show the relevant changes you made?

Reply:
Thanks Levi. I don't want to rely on `CoroutineExceptionHandler` either. The exception handler should be red zone, and ideally never get invoked. But this can pretend as additional layer of protection for preventing crashes in customer app with issues that are hard to identify or we don't expect to happen (e.g. [this one](https://github.com/customerio/customerio-android/issues/155)). And this is mainly, because in KMM, we are doing everything async and not re-creating coroutines every time. I also don't want us to expect our code to crash, but if, by any chance, the code in coroutine throws an exception, the whole coroutine gets stopped until it is recreated.

So, before going for any other solution, I want you and other folks to play around KMM changes a little, to have understanding of the codebase. We can then discuss and plan more workable solution I think. I'm completely okay with dropping CoroutineExceptionHandler, but there should be some alternate for it, that can give slightly more guarantee for preventing crashes and restarting coroutines. 

Right now, I would want to continue with this PR. Take a look at how are we using coroutines and this class in KMM. Then discuss some alternate and go for it if when needed.

For the changes, [this commit](https://github.com/customerio/mobile-shared/pull/5/commits/475e8e8b6c090f745fe76faffa9acea90a58070d) has everything you'll need to see for the mentioned changes.

Reply:
If I am understanding you correctly, what you are proposing is: if code that runs inside of a coroutine throws an exception, our code base should gracefully catch that exception and prevent a customer's app from crashing. So, introducing something like a global exception handler for code that runs inside of a coroutine? Is that correct? 

Assuming that is correct, our SDK can still throw an exception (making an app crash) when code is being executed outside of a coroutine. So, I might not yet agree with this idea? Although I can see value in the idea as an app crash is something we have a goal of avoiding.  

Ideally, I would instead like to use a strategy for how to prevent crashes inside of a coroutine and outside of one. Which, I think leveraging `try/catch` and using the `kotlin.Result` data type would allow us to do? 
 
If we do want to introduce something like a global exception handler, I would want it to be something very simple in our code base. After seeing you removed [the `onFailure` block](https://github.com/customerio/mobile-shared/pull/5/commits/475e8e8b6c090f745fe76faffa9acea90a58070d), the solution you have proposed here thus far I am OK with us going forward with. Seems simple enough to use 👍.

Reply:
> ...something like a global exception handler for code that runs inside of a coroutine?

Correct. Even though every code that runs in coroutine and has any chance of throwing exception should have its own `try/catch`, but global exception handler just adds additional catch to prevent unwanted crashes/errors.

Yes, the code outside can still break. But as we progress more with KMM, we can cover more areas to run code in this or similar blocks to minimize them.

Yes, the strategy should be followed anyway. I too have used `kotlin.runCatching` and `kotlin.Result` at all the places that could throw any exception or where result was helpful.

Glad to hear that. As we develop more features with KMM, we'll see more cases for invoking crashes/errors. We can keep improving this; to look simpler, is easy-to-use and help us minimizing any unexpected crashes/errors.